pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,504 @@
+
+    _logger.info("Number of reduce tasks for pre-computing: {}", numReduceTasks);
+    job.setNumReduceTasks(numReduceTasks);
+
+    // Sort config.
+    if (_sortedColumn != null) {
+      _logger.info("Adding sorted column: {} to job config", _sortedColumn);
+      job.getConfiguration().set(SORTED_COLUMN, _sortedColumn);
+
+      // Sorting is enabled. Adding sorted column value to mapper output key.
+      Schema sortedColumnSchema = schema.getField(_sortedColumn).schema();
+      Schema sortedColumnAsKeySchema;
+      if (sortedColumnSchema.getType().equals(Schema.Type.UNION)) {
+        sortedColumnAsKeySchema = Schema.createUnion(sortedColumnSchema.getTypes());
+      } else {
+        sortedColumnAsKeySchema = Schema.create(sortedColumnSchema.getType());
+      }
+      Schema.Field columnField = new Schema.Field(_sortedColumn, sortedColumnAsKeySchema, "sortedColumn", null);
+      fieldSet.add(columnField);
+    } else {
+      // If sorting is disabled, hashcode will be the only factor for sort/group comparator.
+      addHashCodeField(fieldSet);
+    }
+
+    Schema mapperOutputKeySchema = Schema.createRecord("record", "", "", false);
+    mapperOutputKeySchema.setFields(new ArrayList<>(fieldSet));
+    _logger.info("Mapper output schema: {}", mapperOutputKeySchema);
+
+    AvroJob.setInputKeySchema(job, schema);
+    AvroJob.setMapOutputKeySchema(job, mapperOutputKeySchema);
+    AvroJob.setMapOutputValueSchema(job, schema);
+    AvroJob.setOutputKeySchema(job, schema);
+
+    // Since we aren't extending AbstractHadoopJob, we need to add the jars for the job to
+    // distributed cache ourselves. Take a look at how the addFilesToDistributedCache is
+    // implemented so that you know what it does.
+    _logger.info("HDFS class path: " + _hdfsDefaultClasspathDir);
+    if (_hdfsDefaultClasspathDir != null) {
+      _logger.info("Copying jars locally.");
+      addContainerLocalFiles(job, _hdfsDefaultClasspathDir, job.getConfiguration());
+    } else {
+      _logger.info("Property 'hdfs.default.classpath.dir' not specified.");
+    }
+
+    long startTime = System.currentTimeMillis();
+    // Submit the job for execution.
+    job.waitForCompletion(true);
+    if (!job.isSuccessful()) {
+      throw new RuntimeException("Job failed : " + job);
+    }
+
+    _logger.info("Finished pre-computing job in {}ms", (System.currentTimeMillis() - startTime));
+  }
+
+  @Nullable
+  private TableConfig getTableConfig()
+      throws IOException {
+    try (ControllerRestApi controllerRestApi = getControllerRestApi()) {
+      return controllerRestApi != null ? controllerRestApi.getTableConfig() : null;
+    }
+  }
+
+  /**
+   * Can be overridden to provide custom controller Rest API.
+   */
+  @Nullable
+  private ControllerRestApi getControllerRestApi() {
+    return _pushLocations != null ? new DefaultControllerRestApi(_pushLocations, _rawTableName) : null;
+  }
+
+  private void setDirPermission(Path dirPath)
```

Comment:
i think this is already in open source, right?

Reply:
Another one is in segment creation job, which runs after this pre-processing job.

Reply:
maybe we can put these utility methods somewhere so they can be reused

Reply:
Done.